File: ios/App/App/DibayCallPipPlugin.swift
import Foundation
import Capacitor

/** Native video PiP bridge — delegates to `NativeVideoCallPipPresenter` when flag/runtime active. */
@objc(DibayCallPipPlugin)
public class DibayCallPipPlugin: CAPPlugin, CAPBridgedPlugin {
    static let eventPipModeChanged = "pipModeChanged"
    static let eventPipAction = "pipAction"

    private static weak var pluginInstance: DibayCallPipPlugin?
    private static var lastPipModeByCallId: [String: Bool] = [:]

    public let identifier = "DibayCallPipPlugin"
    public let jsName = "DibayCallPip"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isPipSupported", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "enterCallPip", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "exitCallPip", returnType: CAPPluginReturnPromise),
    ]

    public override func load() {
        super.load()
        DibayCallPipPlugin.pluginInstance = self
    }

    @objc func isPipSupported(_ call: CAPPluginCall) {
        call.resolve(["supported": NativeVideoCallPipPresenter.isSupported()])
    }

    @objc func enterCallPip(_ call: CAPPluginCall) {
        let callId = call.getString("callId") ?? ""
        let ok = NativeVideoCallPipPresenter.requestEnter(callId: callId, source: "plugin")
        call.resolve(["ok": ok])
    }

    @objc func exitCallPip(_ call: CAPPluginCall) {
        let callId = call.getString("callId") ?? ""
        let ok = NativeVideoCallPipPresenter.requestExit(callId: callId)
        call.resolve(["ok": ok])
    }

    static func publishPipModeChanged(inPipMode: Bool, callId: String) {
        let sid = callId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !sid.isEmpty else { return }
        if lastPipModeByCallId[sid] == inPipMode { return }
        lastPipModeByCallId[sid] = inPipMode

        var payload = JSObject()
        payload["inPipMode"] = inPipMode
        payload["callId"] = sid
        emitPipModeChanged(payload)
    }

    static func publishPipAction(action: String, callId: String) {
        let sid = callId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !sid.isEmpty else { return }
        let normalized = action.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard normalized == "restore" || normalized == "end" else { return }

        var payload = JSObject()
        payload["action"] = normalized
        payload["callId"] = sid
        emitPipAction(payload)
    }

    static func clearPipEmitGuards(callId: String) {
        let sid = callId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !sid.isEmpty else { return }
        lastPipModeByCallId.removeValue(forKey: sid)
    }

    private static func emitPipModeChanged(_ payload: JSObject) {
        guard let plugin = pluginInstance else { return }
        plugin.notifyListeners(eventPipModeChanged, data: payload)
    }

    private static func emitPipAction(_ payload: JSObject) {
        guard let plugin = pluginInstance else { return }
        plugin.notifyListeners(eventPipAction, data: payload)
    }
}

File: android/app/src/main/java/com/dibay/app/call/DibayCallPipPlugin.java
package com.dibay.app.call;

import android.os.Build;
import android.content.pm.PackageManager;
import com.dibay.app.MainActivity;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

/** JS bridge — lib/call/native/dibay-call-pip.ts */
@CapacitorPlugin(name = "DibayCallPip")
public class DibayCallPipPlugin extends Plugin {

  private static volatile DibayCallPipPlugin instance;
  private static volatile boolean enteringPip;

  @Override
  public void load() {
    super.load();
    instance = this;
  }

  public static DibayCallPipPlugin getInstance() {
    return instance;
  }

  @PluginMethod
  public void isPipSupported(PluginCall call) {
    boolean supported =
        Build.VERSION.SDK_INT >= Build.VERSION_CODES.O
            && getActivity()
                .getPackageManager()
                .hasSystemFeature(PackageManager.FEATURE_PICTURE_IN_PICTURE);
    JSObject result = new JSObject();
    result.put("supported", supported);
    call.resolve(result);
  }

  @PluginMethod
  public void enterCallPip(PluginCall call) {
    if (enteringPip) {
      JSObject blocked = new JSObject();
      blocked.put("ok", false);
      call.resolve(blocked);
      return;
    }
    if (!(getActivity() instanceof MainActivity)) {
      call.reject("no_main_activity");
      return;
    }
    enteringPip = true;
    try {
      boolean ok = ((MainActivity) getActivity()).requestVideoCallPipFromBridge();
      JSObject result = new JSObject();
      result.put("ok", ok);
      call.resolve(result);
    } finally {
      enteringPip = false;
    }
  }

  @PluginMethod
  public void exitCallPip(PluginCall call) {
    boolean ok = false;
    if (getActivity() != null
        && Build.VERSION.SDK_INT >= Build.VERSION_CODES.N
        && getActivity().isInPictureInPictureMode()) {
      getActivity().moveTaskToBack(false);
      ok = true;
    }
    JSObject result = new JSObject();
    result.put("ok", ok);
    call.resolve(result);
  }

  public void emitPipModeChanged(boolean inPipMode, String callId) {
    JSObject data = new JSObject();
    data.put("inPipMode", inPipMode);
    if (callId != null && !callId.isEmpty()) {
      data.put("callId", callId);
    }
    notifyListeners("pipModeChanged", data);
  }

  public void emitPipAction(String action, String callId) {
    JSObject data = new JSObject();
    data.put("action", action);
    if (callId != null && !callId.isEmpty()) {
      data.put("callId", callId);
    }
    notifyListeners("pipAction", data);
  }

  public void emitPipFallbackDock(String callId) {
    emitPipAction("restore", callId);
  }
}
